x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/ApiKeySearchQueryBuilder.java

```diff
@@ -0,0 +1,157 @@
+            final String translatedFieldName = FieldNameTranslators.translate(query.fieldName());
+            return QueryBuilders.wildcardQuery(translatedFieldName, query.value())
+                .caseInsensitive(query.caseInsensitive())
+                .rewrite(query.rewrite());
+        } else if (qb instanceof RangeQueryBuilder) {
+            final RangeQueryBuilder query = (RangeQueryBuilder) qb;
+            final String translatedFieldName = FieldNameTranslators.translate(query.fieldName());
+            if (query.relation() != null) {
+                throw new IllegalArgumentException("range query with relation is not supported for search");
+            }
+            final RangeQueryBuilder newQuery = QueryBuilders.rangeQuery(translatedFieldName);
+            if (query.format() != null) {
+                newQuery.format(query.format());
+            }
+            if (query.timeZone() != null) {
+                newQuery.timeZone(query.timeZone());
+            }
+            if (query.from() != null) {
+                newQuery.from(query.from()).includeLower(query.includeLower());
+            }
+            if (query.to() != null) {
+                newQuery.to(query.to()).includeUpper(query.includeUpper());
+            }
+            return newQuery.boost(query.boost());
+        } else {
+            throw new IllegalArgumentException("Query type [" + qb.getClass() + "] is not supported for search");
+        }
+    }
+
+    private static class ApiKeyBoolQueryBuilder extends BoolQueryBuilder {
+
+        ApiKeyBoolQueryBuilder(BoolQueryBuilder boolQueryBuilder) {
+            super();
+            minimumShouldMatch(boolQueryBuilder.minimumShouldMatch());
+            adjustPureNegative(boolQueryBuilder.adjustPureNegative());
+            boolQueryBuilder.must().forEach(this::must);
+            boolQueryBuilder.should().forEach(this::should);
+            boolQueryBuilder.mustNot().forEach(this::mustNot);
+            boolQueryBuilder.filter().forEach(this::filter);
+        }
+
+        @Override
+        protected Query doToQuery(SearchExecutionContext context) throws IOException {
+            return super.doToQuery(context.withAllowedFieldNames(ApiKeyBoolQueryBuilder::isFieldNameAllowed));
+        }
+
+        @Override
+        protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+            if (queryRewriteContext instanceof SearchExecutionContext) {
+                return super.doRewrite(
+                    ((SearchExecutionContext) queryRewriteContext).withAllowedFieldNames(ApiKeyBoolQueryBuilder::isFieldNameAllowed)
+                );
+            } else {
+                return super.doRewrite(queryRewriteContext);
+            }
+        }
+
+        static boolean isFieldNameAllowed(String fieldName) {
+            if (Set.of("doc_type", "name", "api_key_invalidated", "creation_time", "expiration_time").contains(fieldName)) {
```

Comment:
this can be extracted in a static variable ?

Reply:
Yep it was done in a rush. Extracted now.